### PR TITLE
Fix forward declaration issues in template/macro context

### LINKED ISSUE #13484: [Macro] NimNode of Forward (Prototype) Procedure Defined with `quote` Cause Error
When we use the forward procedure defined with `quote`, compiler issues following error.

### Example
```nim
proc defForwardQuo(id: NimNode): NimNode =
  result = quote do:
    proc `id`(): int

proc defForwardNew(id: NimNode): NimNode =
  result = newProc(id, @[newIdentNode("int")])

proc def(id: NimNode): NimNode =
  result = quote do:
    proc `id`(): int =
      result = 42

proc call(id: NimNode): NimNode =
  result = quote do:
    `id`()

macro defCallQuo(): untyped =
  # cause an error
  let
    procId = genSym(nskProc)
    protoNode = defForwardQuo(procId)
    defNode = def(procId)
    callNode = call(procId)
  result = quote do:
    `protoNode`
    `defNode`
    echo `callNode`
  echo result.toStrLit

macro defCallNew(): untyped =
  # cause no error
  let
    procId = genSym(nskProc)
    protoNode = defForwardNew(procId)
    defNode = def(procId)
    callNode = call(procId)
  result = quote do:
    `protoNode`
    `defNode`
    echo `callNode`
  echo result.toStrLit

# cause an error
defCallQuo()

# cause no error
# defCallNew()
```

### Current Output
```
proc `:tmp`_141170(): int
proc `:tmp`_141170(): int =
  result = 42

echo `:tmp_141170`()
/.../test.nim(31, 7) Hint: 'defCallNew' is declare
d but not used [XDeclaredButNotUsed]                                             
/.../test.nim(21, 20) Error: internal error: still
 forwarded: :tmp                                                                 
No stack traceback available
To create a stacktrace, rerun compilation with ./koch temp c <file>
```

and the result of `./koch temp c test.nim` is as following

```
proc `:tmp`_185170(): int
proc `:tmp`_185170(): int =
  result = 42

echo `:tmp_185170`()
/.../test.nim(31, 7) Hint: 'defCallNew' is declare
d but not used [XDeclaredButNotUsed]                                             
/.../test.nim(21, 20) Error: internal error: still
 forwarded: :tmp                                                                 
Traceback (most recent call last)
/.../Nim/compiler/nim.nim(118) nim
/.../Nim/compiler/nim.nim(95) handleCmdLine
/.../Nim/compiler/cmdlinehelper.nim(98) loadConfigsAndRunMainC
ommand                                                                           
/.../Nim/compiler/main.nim(188) mainCommand
/.../Nim/compiler/main.nim(95) commandCompileToC
/.../Nim/compiler/cgen.nim(2024) cgenWriteModules
/.../Nim/compiler/cgen.nim(2013) genForwardedProcs
/.../Nim/compiler/msgs.nim(561) internalError
/.../Nim/compiler/msgs.nim(532) liMessage
/.../Nim/compiler/msgs.nim(357) handleError
/.../Nim/compiler/msgs.nim(347) quit
FAILURE
```

### Expected Output
```
proc `:tmp`_185170(): int
proc `:tmp`_185170(): int =
  result = 42

echo `:tmp_185170`()
42
```

### Possible Solution

I have no idea.

### Additional Information

```
$ nim -v
Nim Compiler Version 1.0.6 [MacOSX: amd64]
Compiled at 2020-01-23
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 89b39ee8fe271d0e1b75c15a4c6cf82eb9c13aea
active boot switches: -d:release
```

and

```
$ ./koch temp -v
/.../.nimble/bin/nim c -d:debug --debugger:native -d:nimBetterRun  -v -
d:leanCompiler /.../Nim/compiler/nim                         
Nim Compiler Version 1.0.6 [MacOSX: amd64]
Compiled at 2020-01-23
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 89b39ee8fe271d0e1b75c15a4c6cf82eb9c13aea
active boot switches: -d:release
```

### LINKED ISSUE #15075: Forward declaring proc produces symchoice in template/macro context
This doesn't compile

### Example
```nim
type Cont = ref object of RootObj
  fn*: proc(c: Cont): Cont {.nimcall.}

when true: #Doesn't work
  template genGenTempl(): untyped =
    proc loop(locals: Cont): Cont
    proc loop(locals: Cont): Cont =
      return Cont(fn: loop)
    proc doServer(): Cont =
      return Cont(fn: loop)

  genGenTempl()
else: #This works
  proc loop(locals: Cont): Cont
  proc loop(locals: Cont): Cont =
    return Cont(fn: loop)
  proc doServer(): Cont =
    return Cont(fn: loop)

discard doServer()
```

### Current Output
```
(8, 21) Error: type mismatch: got <None> but expected 'proc (c: Cont): Cont'
```

This error occurs because the `loop` in line 8 is turned into an nkClosedSymChoice that lists the forward declared loop sym and the now defined loop sym as two seperate "overloads".

```
$ nim -v
Nim Compiler Version #devel
```

### PR DESCRIPTION
**Pull request by Clyybber**
Fixes #15075
Fixes #13484

### CODE REVIEW

File: compiler/evaltempl.nim [hunk 1]
```diff
@@ -181,6 +182,7 @@ proc evalTemplate*(n: PNode, tmpl, genSymOwner: PSym;
   ctx.config = conf
   ctx.ic = ic
   initIdTable(ctx.mapping)
+  ctx.gensymHash = rand(0..int.high) #TODO: Wtf
```
Comment (Clyybber) [inline compiler/evaltempl.nim]: @Araq Do we already have a mechanism to get a hash/id thats unique per template invocation or something that could be used for this?

Comment (Araq) [reply compiler/evaltempl.nim]: didn't find anything either. You could add a counter to ``PContext``.

Comment (Clyybber) [reply compiler/evaltempl.nim]: Done, also had to pass it to the VM for `getAst` template expansion

**Comment by timotheecour** [comment]
this causes a regression: https://github.com/nim-lang/Nim/issues/15526 (and same/similar https://github.com/nim-lang/Nim/issues/17162)

**Pull request merged**